### PR TITLE
Improve efficiency of scheduling and token sampiling

### PR DESCRIPTION
**Pull request by tohtana**
(NOTE: This PR requires the new APIs introduced in https://github.com/microsoft/DeepSpeed/pull/4965)

This PR improves the efficiency scheduling for ragged batching.
- Use a new faster API to query KV cache status (94ebae10bae0191f5c3450819285f166f6875bc9)
- Improved efficiency of Top P logits processor by avoiding an inefficient loop (390302476251c9d58ea443900a0388e8d006f19f)
- Skip duplicated check of schedulability (904e5009e2445b830613a8bfb4b5a01634236ab0)
- Use python int values and lists instead of torch tensors to maintain KV cache status. Using torch tensors has a certain overhead and it becomes significant when they are frequently called in loops (bfdb5db019c44df9a642af513ded4a5ac1374be7)

### CODE REVIEW

File: mii/batching/ragged_batching.py [hunk 1]
```diff
@@ -226,12 +225,33 @@ def _generate_output(self, r: Request) -> bool:
         for output in outputs:
             self.result_queues[r.tid].put_nowait(output)
 
-    def _do_schedule_requests(self, requests: List[Request]) -> None:
+    def _schedule_token_gen(self, requests: List[Request]) -> None:
+        free_blocks = min(self.inference_engine.free_blocks)
+        conf_manager = self.inference_engine._config.state_manager
+
+        num_schedulable = min(len(requests), conf_manager.max_ragged_sequence_count)
+        num_schedulable = min(num_schedulable, conf_manager.max_ragged_batch_size)
```
Comment (mrwyattii) [inline mii/batching/ragged_batching.py:233]:
```suggestion
        num_schedulable = min([len(requests), conf_manager.max_ragged_sequence_count, conf_manager.max_ragged_batch_size])
```

File: mii/batching/ragged_batching.py [hunk 2]
```diff
@@ -226,12 +225,33 @@ def _generate_output(self, r: Request) -> bool:
...
+            if block_capacity > 0:
+                self.scheduled_length += 1
+                self.scheduled_requests.append(r)
+            else:
+                # We need a new block
+                if free_blocks > 0:
+                    free_blocks -= 1
+                    self.scheduled_length += 1
+                    self.scheduled_req_blocks += 1
+                    self.scheduled_requests.append(r)
```
Comment (mrwyattii) [inline mii/batching/ragged_batching.py:247]:
```suggestion
            if block_capacity > 0:
                self.scheduled_length += 1
                self.scheduled_requests.append(r)
            elif free_blocks > 0:
                # We need a new block
                free_blocks -= 1
                self.scheduled_length += 1
                self.scheduled_req_blocks += 1
                self.scheduled_requests.append(r)
```

File: requirements/requirements.txt [hunk 3]
```diff
@@ -1,5 +1,5 @@
 asyncio
-deepspeed>=0.12.4
+deepspeed>=0.12.7
```
Comment (mrwyattii) [inline requirements/requirements.txt:2]:
```suggestion
deepspeed>=0.12.4
```

Reverting this, we will update requirements in another PR before next release of DS + MII.

**Review by mrwyattii — APPROVED** [review]
